runtime/parachains/src/paras.rs

```diff
@@ -0,0 +1,1103 @@
+				Ok(_) => T::DbWeight::get().reads_writes(1, 0),
+				Err(i) => {
+					v.insert(i, id);
+					T::DbWeight::get().reads_writes(1, 1)
+				}
+			}
+		})
+	}
+
+	/// Schedule a future code upgrade of the given parachain, to be applied after inclusion
+	/// of a block of the same parachain executed in the context of a relay-chain block
+	/// with number >= `expected_at`
+	///
+	/// If there is already a scheduled code upgrade for the para, this is a no-op.
+	#[allow(unused)]
+	pub(crate) fn schedule_code_upgrade(
+		id: ParaId,
+		new_code: ValidationCode,
+		expected_at: T::BlockNumber,
+	) -> Weight {
+		<Self as Store>::FutureCodeUpgrades::mutate(&id, |up| {
+			if up.is_some() {
+				T::DbWeight::get().reads_writes(1, 0)
+			} else {
+				*up = Some(expected_at);
+				FutureCode::insert(&id, new_code);
+				T::DbWeight::get().reads_writes(1, 2)
+			}
+		})
+	}
+
+	/// Note that a para has progressed to a new head, where the new head was executed in the context
+	/// of a relay-chain block with given number. This will apply pending code upgrades based
+	/// on the block number provided.
+	#[allow(unused)]
+	pub(crate) fn note_new_head(
+		id: ParaId,
+		new_head: HeadData,
+		execution_context: T::BlockNumber,
+	) -> Weight {
+		if let Some(expected_at) = <Self as Store>::FutureCodeUpgrades::get(&id) {
+			Heads::insert(&id, new_head);
+
+			if expected_at <= execution_context {
+				<Self as Store>::FutureCodeUpgrades::remove(&id);
+				let new_code = FutureCode::take(&id);
+
+				let prior_code = CurrentCode::get(&id).unwrap_or_default();
+				CurrentCode::insert(&id, &new_code);
+
+				let now = <system::Module<T>>::block_number();
+
+				let weight = Self::note_past_code(
+					id,
+					expected_at,
+					now,
+					prior_code,
+				);
+
+				// add 1 to writes due to heads update.
+				weight + T::DbWeight::get().reads_writes(3, 1 + 3)
+			} else {
+				T::DbWeight::get().reads_writes(1, 1 + 0)
+			}
+		} else {
+			T::DbWeight::get().reads_writes(1, 0)
+		}
+	}
+
+	/// Fetches the validation code to be used when validating a block in the context of the given
+	/// relay-chain height. A second block number parameter may be used to tell the lookup to proceed
+	/// as if an intermediate parablock has been with the given relay-chain height as its context.
+	/// This may return past, current, or (with certain choices of `assume_intermediate`) future code.
+	///
+	/// `assume_intermediate`, if provided, must be before `at`. If `at` is not within the acceptance
+	/// of the current block number, this will return `None`
+	#[allow(unused)]
+	pub(crate) fn validation_code_at(
+		id: ParaId,
+		at: T::BlockNumber,
+		assume_intermediate: Option<T::BlockNumber>,
+	) -> Option<ValidationCode> {
+		let now = <system::Module<T>>::block_number();
+		let config = <configuration::Module<T>>::config();
+
+		if assume_intermediate.as_ref().map_or(false, |i| &at <= i) {
+			return None;
+		}
+
+		if at + config.acceptance_period + One::one() < now {
```

Comment:
Either the `+ One::one()` is on the wrong side of the comparison to emulate a `<=` or it needs some additional explanation why the `acceptance` period is prolonged by one.
>  `assume_intermediate`, if provided, must be before `at`. If `at` is not within `config.acceptance_period` of the current block number, this will return None.
```latex
ap := config.acceptance_period
at \in [now - ap .. now] 
now - ap <= at
inversion / abort criteria:
! (now - ap  <= at )
now - ap > at
now > at + ap
```

Reply:
TBH, I'm not sure why I added this constraint anymore. The goal was to prevent doing `validation_code_at` requests for stuff that has been pruned, but the `PastCodeMeta::code_at` function can safely return `None` when the data has been pruned, so it doesn't seem necessary.

Furthermore the constraint may actually lead to us rejecting validation code requests where we still have the code, in cases like `(10, 15)`, and `now = 10 + acceptance_period`. This is clearly bad, as we care when `15` exits the acceptance period, not when `10` does.

So I'll remove this constraint and alter the tests to ensure that those cases are covered

Reply:
ffe66b136a7d